content/augmentor.js

```diff
@@ -0,0 +1,198 @@
+
+// You should have received a copy of the GNU General Public License
+// along with Parity.  If not, see <http://www.gnu.org/licenses/>.
+
+import { h, render } from 'preact';
+/** @jsx h */
+
+import { AugmentedIcon } from './components';
+
+import Runner from './runner';
+import { FETCH_IMAGE } from '../background/processor';
+
+import styles from './styles.less';
+
+export const AUGMENTED_NODE_ATTRIBUTE = 'data-parity-touched';
+
+export default class Augmentor {
+
+  static run (matches, resolved = {}) {
+    // Use the attributes matcher first
+    const attributesMatches = matches.filter((match) => match.from === 'attributes');
+    const textMatches = matches.filter((match) => match.from === 'text');
+
+    attributesMatches
+      .forEach((match) => {
+        const { email, name, node } = match;
+        const value = email || name;
+
+        Augmentor.augmentNode(value, node, resolved);
+      });
+
+    textMatches
+      .forEach((match) => {
+        const { email, name } = match;
+        const value = email || name;
+        const safeNode = Augmentor.getSafeNode(match);
+
+        Augmentor.augmentNode(value, safeNode, resolved);
+      });
+  }
+
+  static getSafeNode (match) {
+    const { email, node } = match;
+
+    // Safe Node is if the node which inner text is only the email address
+    if (node.innerText.trim() === email) {
+      return node;
+    }
+
+    const emailIndex = node.innerText.indexOf(email);
+
+    if (emailIndex === -1) {
+      return;
+    }
+
+    const beforeText = node.innerText.slice(0, emailIndex);
+    const afterText = node.innerText.slice(emailIndex + email.length);
+
+    const safeNode = document.createElement('span');
+    safeNode.innerText = email;
+
+    const nextNode = node.cloneNode(true);
+    nextNode.innerHTML = '';
+    nextNode.appendChild(safeNode);
+    safeNode.insertAdjacentText('beforebegin', beforeText);
+    safeNode.insertAdjacentText('afterend', afterText);
+
+    // Replace the node with the safe node
+    node.parentElement.replaceChild(nextNode, node);
+
+    return safeNode;
+  }
+
+  static augmentNode (key, node, resolved = {}) {
+    if (!node || node.getAttribute(AUGMENTED_NODE_ATTRIBUTE) === 'true') {
+      return;
+    }
+
+    const data = resolved[key];
+    node.setAttribute(AUGMENTED_NODE_ATTRIBUTE, true);
+
+    if (!data) {
+      return;
+    }
+
+    return Augmentor.fetchImages(data)
+      .then(([ badges, tokens ]) => {
+        const { address, name } = data;
+        const { height = 16 } = node.getBoundingClientRect();
+
+        const augmentedIcon = render((
+          <AugmentedIcon
+            address={ address }
+            badges={ badges }
+            height={ height }
+            name={ name }
+            tokens={ tokens }
+          />
+        ));
+
+        const container = document.createElement('span');
+        container.className = styles.container;
+        container.appendChild(node.cloneNode(true));
```

Comment:
I think it would be better to avoid cloning and `appendChild` after we replace it in `parentElement`, that way we don't remove any listeners that might be attached to this node.